cmake_modules/FindSnappy.cmake

```diff
@@ -11,57 +10,78 @@
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 # See the License for the specific language governing permissions and
 # limitations under the License.
+#
+# Tries to find Snappy headers and libraries.
+#
+# Usage of this module as follows:
+#
+#  find_package(Snappy)
+#
+# Variables used by this module, they can change the default behaviour and need
+# to be set before calling find_package:
+#
+#  Snappy_HOME - The root of the Snappy installation.  This is considered before
+#                standard locations. The environment variable SNAPPY_HOME is also
+#                considered if this variable is not set.
+#  Snappy_USE_STATIC_LIBS - Set to ON to force the use of the static
+#                           libraries.  Default is OFF
+#
+# Variables defined by this module:
+#
+#  SNAPPY_FOUND              System has Snappy libs/headers
+#  SNAPPY_LIBRARIES          The Snappy libraries
+#  SNAPPY_INCLUDE_DIR        The location of Snappy headers
+
+# Determine parameterized roots
+if( Snappy_HOME )
+    list( APPEND _snappy_roots ${Snappy_HOME} )
+endif()
+if( NOT "$ENV{SNAPPY_HOME}" STREQUAL "")
+    file( TO_CMAKE_PATH "$ENV{SNAPPY_HOME}" _native_path )
+    list( APPEND _snappy_roots ${_native_path} )
+endif()
 
-# - Find SNAPPY (snappy.h, libsnappy.a, libsnappy.so, and libsnappy.so.1)
-# This module defines
-#  SNAPPY_INCLUDE_DIR, directory containing headers
-#  SNAPPY_LIBS, directory containing snappy libraries
-#  SNAPPY_STATIC_LIB, path to libsnappy.a
-#  SNAPPY_FOUND, whether snappy has been found
+# Try the parameterized roots first
+find_path( SNAPPY_INCLUDE_DIR NAMES snappy.h
+    PATHS ${_snappy_roots} NO_DEFAULT_PATH
+    PATH_SUFFIXES "include" )
+find_path( SNAPPY_INCLUDE_DIR NAMES snappy.h )
 
-set(SNAPPY_SEARCH_HEADER_PATHS
-  ${THIRDPARTY_PREFIX}/include
-)
+# Support preference of static libs by adjusting CMAKE_FIND_LIBRARY_SUFFIXES.
+# Patterned after FindBoost.cmake
+if( Snappy_USE_STATIC_LIBS )
```

Comment:
When I was researching this, I found a [limitation](https://cmake.org/Bug/view.php?id=1643) that cmake's default `find_library` behavior prefer dynamic libraries.  The original code finds a dynamic library and then assumes that the static library exists in the same path.  I didn't think this was a strong enough guarantee.

To ensure you can find both, you have to make two calls to `find_library`, once looking for dynamic and once looking for static.  Rather than always make the two call, since a particular use of FindSnappy.cmake may only care about one, it is modal such that you can call it indicating which type of library you care about.  So you can actually call `find_package( Snappy ... )` once without `Snappy_USE_STATIC_LIBS` set and once with it set to get both libraries.  I figured since this is what the standard FindBoost.cmake did, then it was best practice for ensuring the static libraries exist.  If the assumption route is good enough then I can pull out the modal stuff, but I felt this could support all the use cases with a stronger guarantee.

As for the library organization stuff... I'm not very knowledgeable here.  I've had several questions myself.  For instance since libparquet uses snappy and lz4 internally and not in an api, it seems reasonable that you could release a library of libparquet.a which doesn't have dependencies.  However if executable is using libparquet.a and libsnappy.a, is there a chance that we'd have duplicate symbols when trying to link?  Can you hide the snappy symbols that libparquet uses so that only libparquet can see them?  Or do you have to ensure that both the binary and librparquet.a are using the same libsnappy.a and only link it once?  On one side you might have duplicate code that is making your binary larger.  On the other hand you lock both the binary and libparquet.a into using the same version of libsnappy.

It's hard to know what to search for, but it seems to me, with thing like how pkg-config operate, that one typically follows the second option, ensuring that both binary and libparquet are compiled against the same libsnappy library and only link it once.